src/actions/receive-profile.js

```diff
@@ -390,10 +394,43 @@ export function finalizeActiveTabProfileView(
       localTrackOrderByPid,
       activeTabHiddenGlobalTracksGetter,
       activeTabHiddenLocalTracksByPidGetter,
+      showTabOnly,
     });
   };
 }
 
+/**
+ * Re-compute the profile view data. That's used to be able to switch between
+ * full and active tab view.
+ */
+export function changeViewAndRecomputeProfileData(
+  showTabOnly: BrowsingContextID | null
+): ThunkAction<void> {
+  return (dispatch, getState) => {
+    const profile = getProfile(getState());
+    // We are resetting the selected thread index, because we are not sure if
+    // the selected thread will be availabe in the next view.
+    const selectedThreadIndex = 0;
+    dispatch({
+      type: 'DATA_RELOAD',
+    });
+
+    if (showTabOnly === null) {
+      // The url state says this is a full view. We should compute and initialize
+      // the state relevant to that state.
+      dispatch(
+        finalizeFullProfileView(profile, selectedThreadIndex, showTabOnly)
```

Comment:
I'm not sure how this is used in the bigger picture, but it would probably be good to retain an already computed `fullProfileView`, and only switch between them when changing views. This could be a follow-up, or perhaps you already have it planned. Perhaps the entire profile view state could be duplicated as needed. Again, I don't have a huge amount of context here, just listing out some of my thoughts.

Action needed: Either disagree with me, or we should file an issue to follow-up. Maybe we can chat tomorrow.

Reply:
Thinking through some ideas on how to structure this...

```ts
type ProfileView = {
  commonProperty1: CommonProperty1,
  commonProperty2: CommonProperty1,
}

type ActiveTabProfileView = {
  customProp1: ActiveTabCustomProp1,
  customProp2: ActiveTabCustomProp2,
  profileView: ProfileView,
};

type FullProfileView = {
  customPropA: FullCustomPropA,
  customPropB: FullCustomPropB,
  profileView: ProfileView,
};

type ProfileViews = {
  activeTab: null | ActiveTabProfileView,
  full: null | FullProfileView,
}

const getHasActiveTabProfileView: Selector<boolean> = state => Boolean(state.profileViews.activeTab);
const getHasFullProfileView: Selector<boolean> = state => Boolean(state.profileViews.full);

/**
 * Get the profile view associated either the full or active tab view.
 */
const getProfileView: Selector<ProfileView> = createSelector(
  getShowTabOnly,
  getGrofileViews,
  (showTabOnly, profileViews) =>
    ensureExists(showTabOnly ? state.profileViews.activeTab.profileView : state.profileViews.full.profileView)
);

export function changeView(
  showTabOnly: BrowsingContextID | null
): ThunkAction<void> {
  return (dispatch, getState) => {
    const profile = getProfile(getState());
    if (showTabOnly === null) {
      if (getHasActiveTabProfileView(getState())) {
        dispatch(switchToActiveTab());
      } else {
        dispatch(
          finalizeFullProfileView(profile, selectedThreadIndex, showTabOnly)
        );
      }
    } else {
      // Do the same for full profile.
    }
  };
}
```

Reply:
Or maybe

```ts
type ProfileView = {
  commonProperty1: CommonProperty1,
  commonProperty2: CommonProperty1,
}

type ActiveTabProfileView = {
  customProp1: ActiveTabCustomProp1,
  customProp2: ActiveTabCustomProp2,
  cachedProfileView: null | ProfileView,
};

type FullProfileView = {
  customPropA: FullCustomPropA,
  customPropB: FullCustomPropB,
  cachedProfileView: null | ProfileView,
};

type ProfileViews = {
  activeTab: null | ActiveTabProfileView,
  full: null | FullProfileView,
  profileView: ProfileView
}
```

Reply:
Yes, I wanted to make a cached state so we can actually skip computing all the time. I initially wanted to make it something like you mentioned. Tried this for example:
```
type ProfileViewState = {
  ...commonProps
  activeTab: null | ActiveTabProfileView,
  full: null | FullProfileView,
}
```
Selectors are working fine with this but the problem with that is the reducers. `combineReducers` doesn't allow you to pass null instead of an object. And that makes this solution not possible. It looks like redux doesn't want you to create optional reducers.

For example the flow error I get when I do a state object nullable:
```
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/types/libdef/npm/redux_v3.x.x.js:56:87

null [1] is incompatible with object type [2] in property activeTabProfile.

     src/types/libdef/npm/redux_v3.x.x.js
     53│   declare export function bindActionCreators<A, C: ActionCreator<A, any>, D: DispatchAPI<A>>(actionCreator: C, dispatch: D): C;
     54│   declare export function bindActionCreators<A, K, C: ActionCreators<K, A>, D: DispatchAPI<A>>(actionCreators: C, dispatch: D): C;
     55│
 [2] 56│   declare export function combineReducers<O: Object, A>(reducers: O): CombinedReducer<$ObjMap<O, <S>(r: Reducer<S, any>) => S>, A>;
     57│
     58│   declare export var compose: $Compose;
     59│ }

     src/types/state.js
 [1] 93│   +activeTabProfile: null | ActiveTabProfileViewState,
```
or this error if you make it an empty object instead of null:
```
Property hiddenGlobalTracksGetter is missing in object type [1] but exists in object type [2] in property
activeTabProfile.

     src/types/libdef/npm/redux_v3.x.x.js
     53│   declare export function bindActionCreators<A, C: ActionCreator<A, any>, D: DispatchAPI<A>>(actionCreator: C, dispatch: D): C;
     54│   declare export function bindActionCreators<A, K, C: ActionCreators<K, A>, D: DispatchAPI<A>>(actionCreators: C, dispatch: D): C;
     55│
 [2] 56│   declare export function combineReducers<O: Object, A>(reducers: O): CombinedReducer<$ObjMap<O, <S>(r: Reducer<S, any>) => S>, A>;
     57│
     58│   declare export var compose: $Compose;
     59│ }

     src/types/state.js
 [1] 93│   +activeTabProfile: {||} | ActiveTabProfileViewState,
```
I tried to search for it but couldn't find a better solution. But maybe there is and I just don't know it yet. Do you have any recommendations on this?

Reply:
Created #2485 after our discussion.